rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs

```diff
@@ -0,0 +1,1089 @@
+                LastIndexedSnapshot {
+                    sequence: Some(6),
+                    at_block: 100,
+                }
+            );
+        }
+    }
+
+    mod sequence_range {
+        use super::*;
+
+        const INDEX_MODE: IndexMode = IndexMode::Sequence;
+        const CHUNK_SIZE: u32 = 10;
+
+        async fn get_cursor() -> ForwardSequenceAwareSyncCursor<MockSequencedData> {
+            let mut cursor = get_test_forward_sequence_aware_sync_cursor(INDEX_MODE, CHUNK_SIZE);
+            // Fast forwarded to sequence 5, block 90
+            cursor.fast_forward().await.unwrap();
+
+            cursor
+        }
+
+        /// Tests successful fast forwarding & successful indexing with a correct sequence range.
+        #[tracing_test::traced_test]
+        #[tokio::test]
+        async fn test_normal_indexing() {
+            let mut cursor = get_cursor().await;
+
+            // We should have fast forwarded to sequence 5, block 90
+            assert_eq!(
+                cursor.current_indexing_snapshot,
+                TargetSnapshot {
+                    sequence: 5,
+                    at_block: 90,
+                }
+            );
+
+            // As the latest sequence count is 5 and the current indexing snapshot is sequence 5, we should
+            // expect no range to index.
+            let range = cursor.get_next_range().await.unwrap();
+            assert_eq!(range, None);
+
+            // Update the tip, expect to still not index anything.
+            cursor.latest_sequence_querier = Arc::new(MockLatestSequenceQuerier {
+                latest_sequence_count: Some(5),
+                tip: 110,
+            });
+            let range = cursor.get_next_range().await.unwrap();
+            assert_eq!(range, None);
+
+            // Update the latest sequence count to 6, now we expect to index.
+            cursor.latest_sequence_querier = Arc::new(MockLatestSequenceQuerier {
+                latest_sequence_count: Some(6),
+                tip: 120,
+            });
+
+            // Expect the range to be:
+            // (new sequence, new sequence)
+            let range = cursor.get_next_range().await.unwrap().unwrap();
+            let expected_range = 5..=5;
+            assert_eq!(range, expected_range);
+
+            // Expect the target snapshot to be set to the latest sequence and tip.
+            assert_eq!(
+                cursor.target_snapshot,
+                Some(TargetSnapshot {
+                    sequence: 5,
+                    at_block: 120,
+                })
+            );
+
+            // Getting the range again without updating the cursor should yield the same range.
+            let range = cursor.get_next_range().await.unwrap().unwrap();
+            assert_eq!(range, expected_range);
+
+            // Update the cursor with the found log.
+            cursor
+                .update(
+                    vec![(MockSequencedData::new(5), log_meta_with_block(115))],
```

Comment:
If I set the return value of `latest_sequence_querier` above to `sequence: 5, at_block: 88` (updating the target snapshot assertion as well), this call to `update` still works, which feels odd. Should we also filter out entries with `block > target_snapshot.block` in `update`?

Reply:
doesn't setting the latest sequence count to 5 result in a None action? or did you mean sequence: 6

imo `update` can assume that any logs passed into the cursor are legitimate, and it should succeed as long as it builds upon the last indexed snapshot and there are no gaps in the logs

Reply:
Indeed I meant sequence 6. What I mean is - what if there's a reorg? 

 You'd end up with a `TargetSnapshot { seq: 6, block: 88 }`, and `LastIndexedSnapshot { seq: 5, block: 115 }`, but also with a `get_next_range` = `110..=88`

Reply:
mm good question

I think it's probably okay to not worry about logic to accommodate reorgs. The `latest_sequence_count_and_tip` is meant to return only the finalized sequence count & the finalized tip. So as long as we've configured the node with a correct reorg period it should never encounter a reorg, so imo we can assume that all logs that make it to the `update` function are finalized. If the agent is misconfigured then I think it's ok to have undefined behavior - but certainly adding logs in such a case to complain loudly makes sense to me, I'll add some